pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,664 @@
+
+    # A temp folder for making a package copy in case it's a local folder,
+    # because extracting metadata might modify files
+    # (creating sdists/wheels...)
+    temp_folder = tempfile.mkdtemp(prefix="pythonpackage-package-copy-")
+    try:
+        # Package is indeed a folder! Get a temp copy to work on:
+        if is_filesystem_path(package):
+            shutil.copytree(
+                parse_as_folder_reference(package),
+                os.path.join(temp_folder, "package")
+            )
+            package = os.path.join(temp_folder, "package")
+
+        # Because PEP517 can be noisy and contextlib.redirect_* fails to
+        # contain it, we will run the actual analysis in a separate process:
+        try:
+            subprocess.check_output([
+                sys.executable,
+                "-c",
+                "import importlib\n"
+                "import json\n"
+                "import os\n"
+                "import sys\n"
+                "sys.path = [os.path.dirname(sys.argv[3])] + sys.path\n"
+                "m = importlib.import_module(\n"
+                "    os.path.basename(sys.argv[3]).partition('.')[0]\n"
+                ")\n"
+                "m._extract_metainfo_files_from_package_unsafe("
+                "    sys.argv[1],"
+                "    sys.argv[2],"
+                ")",
+                package, output_folder, os.path.abspath(__file__)],
+                stderr=subprocess.STDOUT,  # make sure stderr is muted.
+                cwd=os.path.join(os.path.dirname(__file__), "..")
+            )
+        except subprocess.CalledProcessError as e:
+            output = e.output
+            try:
+                output = output.decode('utf-8', 'replace')
+            except AttributeError:
+                pass
+            if debug:
+                print("Got error obtaining meta info.")
+                print("Detail output:")
+                print(output)
+                print("End of Detail output.")
+            raise ValueError(
+                "failed to obtain meta info - "
+                "is '{}' a valid package? "
+                "Detailed output:\n{}".format(package, output)
+                )
+    finally:
+        shutil.rmtree(temp_folder)
+
+
+def _get_system_python_executable():
+    """ Returns the path the system-wide python binary.
+        (In case we're running in a virtualenv or venv)
+    """
+    # This function is required by get_package_as_folder() to work
+    # inside a virtualenv, since venv creation will fail with
+    # the virtualenv's local python binary.
+    # (venv/virtualenv incompatibility)
+
+    # Abort if not in virtualenv or venv:
+    if not hasattr(sys, "real_prefix") and (
+            not hasattr(sys, "base_prefix") or
+            os.path.normpath(sys.base_prefix) ==
+            os.path.normpath(sys.prefix)):
+        return sys.executable
+
+    # Extract prefix we need to look in:
+    if hasattr(sys, "real_prefix"):
```

Comment:
looks fishy, but I'll take it

Reply:
The meaning of these is explained here: https://stackoverflow.com/questions/1871549/determine-if-python-is-running-inside-virtualenv

Reply:
Thanks! Yes I was assuming something like this :smiley: 
That's usually how I end up writing some one-liner methods :joy: because it's can be more self explanatory